Cleaned up Azure::Core::Context API surface (#5676)
Azure/azure-sdk-for-cpp: README.md

  std::string sourceUri = "<a uri to the source blob to copy>";

>   // Typically, long running operation APIs have names that begin with Start.

Reviewer: Odd to keep the comment w/o an example.

Author: Overaggressive cut and paste (I accidentally put the context example in the LRO example and accidentally removed a line.

  }
```

#### `Azure::Core::Context`

Most Azure SDK Service Client methods accept an optional `Azure::Core::Context` parameter, which is used to enable cancellation of the operation.

This is useful when you want to cancel an operation that is taking too long to complete. For instance, the 
snippet below will cancel a blob client upload after 500 milliseconds.

```cpp
  Azure::Core::Context cancelledIn500ms = Azure::Core::Context::ApplicationContext.WithDeadline(
          std::chrono::system_clock::now() + std::chrono::milliseconds(500));

  std::cout << "Uploading blob: " << blobName << std::endl;
  blobClient.UploadFrom(blobContent, sizeof(blobContent), cancelledIn500ms);
```

`Context` objects can also be directly cancelled using the `Context::Cancel` method.

> `Context` objects form a hierarchy, where a child context can created from a parent context. 

Reviewer: ```suggestion
`Context` objects form a hierarchy, where a child context can be created from a parent context. 
```

There are five operations that can be performed on a 'Context' object:

* Create a child context from a parent context which can be independently cancelled.
* Create a child context from a parent context with a Key/Value pair. This is useful for associating metadata with a context.

> * Create a child context from a parent context with a deadline. This is useful for setting a deadline for a context.

Reviewer: ```suggestion
* Create a child context from a parent context with a deadline. This is useful for setting a timeout.
```

  }
```

#### `Azure::Core::Context`

> Most Azure SDK Service Client methods accept an optional `Azure::Core::Context` parameter, which is used to enable cancellation of the operation or to

Author: Most.

Specifically, all "service methods" accept a `Context` parameter. But not all client methods are service methods.

As a degenerate example, the constructor does not take a Context parameter, because the constructor is not a service method.

But while the Azure SDK team has a distinguished category of APIs known as "service methods", we don't typically expose it as a discrete category of APIs to customers.

establish an absolute deadline for the operation.

This is useful when you want to assign a time limit on an operation to ensure that it completes in a "reasonable" timeframe. For instance, the 
snippet below will cancel a blob client upload after 5 seconds.

```cpp
  Azure::Core::Context cancelledIn5s = Azure::Core::Context::ApplicationContext.WithDeadline(
          std::chrono::system_clock::now() + std::chrono::seconds(5);

  std::cout << "Uploading blob: " << blobName << std::endl;
  try {
    blobClient.UploadFrom(blobContent, sizeof(blobContent), cancelledIn5s);
  }
  catch (const Azure::Core::OperationCancelledException& e) {
	std::cout << "Upload was cancelled: " << e.what() << std::endl;

>   }

Author: That is normal for snippets. Snippets are not full programs, they are small chunks of code which are used to demonstrate an example.

One core aspect of a snippet is that it is concise and only includes the elements which are absolutely required for the snippet.

* Create a child context from a parent context which can be independently cancelled.
* Create a child context from a parent context with a Key/Value pair. This is useful for associating metadata with a context.
* Create a child context from a parent context with a deadline. This is useful for setting a timeout.
* Cancel a context. This will cancel the context and all its children. Note that there is no way of un-cancelling a context.
* Check if a context is cancelled.

By default, when a `Context` is created, it is a child of the `ApplicationContext`. This allows a client application to cancel all the operations currently active in the process.

If you do NOT want a newly created `Context` to be a child of `ApplicationContext`, use the `Azure::Core::Context::CreateNewRoot()` function, 
which will create a new root context.

When a context is copied from another context, the copied context will share state with the original context. 
This means that if the original context is cancelled, the copied context will also be cancelled.

Cancellation of a `Context` is a permanent operation. Once a non-root context is cancelled, it cannot be un-cancelled.

> Root `Contexts` are unique in that their cancellation *CAN* be reset. Cancelling a root context essentially creates a new root Context replacing the existing root.

Author: The semantic distinction between Root and non-root contexts is implicit in the design for Context. Whenever you call "WithDeadline", "WithValue", or "WithContext", you are explicitly creating a parent/child relationship between contexts. And that defines a tree. 

And every tree needs a root.

There is no such thing as a "Regular" context.

All Context object have a parent object. The core aspect of this change is that in previous versions, the relationship between a newly created Context object and ApplicationContext was described in the documentation but not in the implementation. 

This change makes the relationship described in the documentation explicit.

And there is basically no situation where an end-user would want to go down that path. They simply don't need to know these details.

What an end-user needs to know is:
1) Every API takes a Context object.
1) If you want to cancel a call, cancel the Context for the call.
1) If you want to cancel ALL calls in the process, cancel ApplicationContext.
    1) If you have cancelled ApplicationContext, you can call ApplicationContext::Reset to start new operations, otherwise your application is toast.

The `CreateNewRoot` method exists for two reasons:

1) It allows us to create ApplicationContext, and
2) It allows a customer who wants the old behavior back to implement the functionality from before the change.

```cpp
  // Create a child context from the root context
  Azure::Core::Context childContext = Azure::Core::Context::ApplicationContext.WithDeadline(std::chrono::system_clock::now() + std::chrono::seconds(10));

  // Check if the child context is cancelled - this will not be printed unless 10 seconds have passed from the creation of childContext.
  if (childContext.IsCancelled())
  {
	std::cout << "Child context is cancelled" << std::endl;
  }

  // Cancel the root context
  Azure::Core::Context::ApplicationContext.Cancel();

  // Check if the child context is cancelled - this will be printed because the root context was cancelled.
  if (childContext.IsCancelled())
  {
	std::cout << "Child context is cancelled" << std::endl;
  }

  // Reset the root context
  Azure::Core::Context::ApplicationContext.Reset();

  // Check the child context. This will be printed because even though the root context was reset, existing contexts created before
  // the reset will still be cancelled.
  if (childContext.IsCancelled())
  {
    std::cout << "Child context is still cancelled" << std::endl;
  }
```

> When a `Context` is cancelled, it is typically indicated by throwing `Azure::Core::OperationCancelledException`. This exception can be caught and handled by the application.

Reviewer 2: Thoughts on this suggestion?

#### Public, Private and Internal Types

> The APIs defined in the Azure SDK for C++ fall into three rough categories:

Author: The SDK API surface is sufficiently complicated that I'd rather use weasel-words to give us some flexibility. Without "rough", then we potentially can paint ourselves in a corner when we run into something that is in the boundaries - in this case, I'm particularly concerned about internal/detail namespaces.

- **Public**: These are the types that are intended to be used by the consumers of the SDK.
They are part of the public API and are stable. Breaking changes to these types will be avoided as much
as possible. All public types and functions are located are in the `Azure` namespace.
- **Internal**: These are the types that are used internally by the packages within the SDK.
They are intended for use by the packages which make up the Azure SDK. They are NOT intended to be used by the consumers of the SDK.

> Breaking changes to these types are allowed, within certain constraints. These types are located in an `Azure` namespace within the `_internal` terminal namespace (for instance, `Azure::Core::Http::Policies::_internal::RequestActivityPolicy`).

Author: I use visual studio code, and it likes to reformat lines.

Also, I have this annoying tendency of wanting to be able to read all the text in the document, so I tend to reformat text so that the words fit on my screen.

- **Private**: These are the types that are used internally to individual Azure SDK Packages.
They are not intended to be used by the consumers of the SDK, or by other SDK packages. Breaking changes
to these types are allowed. These types are located in an `Azure` namespace within the `_detail` terminal namespace.

> Within the source tree, Internal types are typically declared in directory named "internal", and Private types are typically declared in a directory named "private".

Author: This part is a rough guideline.

We should define explicit rules and put them in the API guidelines (which is a different pull request outstanding).

  }
```

#### `Azure::Core::Context`

Most Azure SDK Service Client methods accept an optional `Azure::Core::Context` parameter, which is used to enable cancellation of the operation or to
establish an absolute deadline for the operation.

This is useful when you want to assign a time limit on an operation to ensure that it completes in a "reasonable" timeframe. For instance, the 
snippet below will cancel a blob client upload after 5 seconds.

<!-- @insert_snippet: CreateBlobContext -->
```cpp

>       Azure::Core::Context cancelledIn5s = Azure::Core::Context::CreateWithDeadline(

Reviewer 3: BTW, do we still have this `CreateWithDeadline()` method, I can't seem to find it.

Reviewer 2: No, we don't. It's a ctor overload, so we should update this sample snippet.

>       Azure::Core::Context cancelledIn5s = Azure::Core::Context::CreateWithDeadline(

Reviewer 2: The sample should be updated to use the ctor overload.

>           std::chrono::system_clock::now() + std::chrono::seconds(5));

Author: If we do that, we need to remove the constructor that takes a deadline, since we want there to be only one way of doing things.

But upon reflection, I think that it's a better experience - less magic.

      auto containerClient = BlobContainerClient::CreateFromConnectionString(
          GetConnectionString(), containerName + std::to_string(i));

>       containerClient.CreateIfNotExists({}, cancelledIn5s);

Reviewer 2: Rnrelated: I wish we had a better dev exp when we have multiple optional parameters in C++. Having to pass "{}" for the options, to be able to provide context is a bit challenging to read.

Author: Unfortunately that's not the way C++ works. The only optional parameters are the *last* parameters in the method. And there's a parameter *after* the options parameter for CreateIfNotExists.

      for (int j = 0; j < 3; ++j)
      {
        BlockBlobClient blobClient
            = containerClient.GetBlockBlobClient(blobName + std::to_string(j));
        blobClient.UploadFrom(
            reinterpret_cast<const uint8_t*>(blobContent.data()),
            blobContent.size(),
            {},
            cancelledIn5s);

>       }

Reviewer 2: Remove the first level indentation

Author: No.

```

`Context` objects can also be directly cancelled using the `Cancel()` method.

`Context` objects form a directed tree, where a child context can be created from a parent context. 
The context tree is unidirectional and acyclic.

These are the basic operations that can be performed on a `Context` object:

* Create a child context from a parent context with a Key/Value pair. This is useful for associating metadata with a context.
* Create a child context from a parent context with a deadline. This is useful for setting a timeout.
* Cancel a context. This will cancel the context and all its children. Note that there is no way of un-cancelling a context.
* Check if a context is cancelled.

When a context is copied from another context, the copied context will share state with the original context. 
This means that if the original context is cancelled, the copied context will also be cancelled.

Cancellation of a `Context` is a permanent operation. Once a context is cancelled, it cannot be un-cancelled.

> When a `Context` is cancelled, it is typically indicated by throwing `Azure::Core::OperationCancelledException`. This exception can be caught and handled by the application.

Reviewer 2: https://github.com/Azure/azure-sdk-for-cpp/pull/5676#discussion_r1654054423
Bringing up an unresolved comment (it was hidden in the previous comments):

nit: The wording makes it seem like calling Cancel on Context will throw a `OperationCancelledException`

```suggestion
When a client operation fails with an `Azure::Core::OperationCancelledException`, it is typically due to a `Context` getting cancelled. This exception can be caught and handled by the application.
```

  }
```

#### `Azure::Core::Context`

Most Azure SDK Service Client methods accept an optional `Azure::Core::Context` parameter, which is used to enable cancellation of the operation or to
establish an absolute deadline for the operation.

This is useful when you want to assign a time limit on an operation to ensure that it completes in a "reasonable" timeframe. For instance, the 
snippet below will cancel a blob client upload after 500 milliseconds.

```cpp
  Azure::Core::Context cancelledIn500ms = Azure::Core::Context::ApplicationContext.WithDeadline(

>           std::chrono::system_clock::now() + std::chrono::milliseconds(500));

Author: Honestly, I would have preferred a `WithTimeout` method rather than `WithDeadline`. But that ship sailed a long time ago.

And I'm a firm believer that there should only be a single way of doing things, so while I like the idea of `WithTimeout`, we already have a `WithDeadline` which isn't *that* worse an experience.

>           std::chrono::system_clock::now() + std::chrono::milliseconds(500));

Author: :). I intentionally chose 500ms as a "silly" value.

But in the light of "snippets are templates", I'll pick a better number.

This project may contain trademarks or logos for projects, products, or services. Authorized use of Microsoft trademarks or logos is subject to and must follow [Microsoft's Trademark & Brand Guidelines](https://www.microsoft.com/legal/intellectualproperty/trademarks/usage/general). Use of Microsoft trademarks or logos in modified versions of this project must not cause confusion or imply Microsoft sponsorship. Any use of third-party trademarks or logos are subject to those third-party's policies.

![Impressions](https://azure-sdk-impressions.azurewebsites.net/api/impressions/azure-sdk-for-cpp%2FREADME.png)

> 

Reviewer 2: nit: remove?
```suggestion
```
